libs/vkd3d/device.c

```diff
@@ -4311,18 +4312,136 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreateSharedHandle(d3d12_device_if
         ID3D12DeviceChild *object, const SECURITY_ATTRIBUTES *attributes, DWORD access,
         const WCHAR *name, HANDLE *handle)
 {
-    FIXME("iface %p, object %p, attributes %p, access %#x, name %s, handle %p stub!\n",
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    struct d3d12_device *device = impl_from_ID3D12Device(iface);
+    ID3D12Resource *resource_iface;
+
+    TRACE("iface %p, object %p, attributes %p, access %#x, name %s, handle %p\n",
             iface, object, attributes, access, debugstr_w(name), handle);
 
+    if (SUCCEEDED(ID3D12DeviceChild_QueryInterface(object, &IID_ID3D12Resource, (void**)&resource_iface)))
+    {
+        struct d3d12_resource *resource = impl_from_ID3D12Resource(resource_iface);
+        VkMemoryGetWin32HandleInfoKHR win32_handle_info;
+        VkResult vr;
+
+        if (!(resource->heap_flags & D3D12_HEAP_FLAG_SHARED))
+        {
+            ID3D12Resource_Release(resource_iface);
+            return DXGI_ERROR_INVALID_CALL;
+        }
+
+        if (attributes)
+            FIXME("attributes %p not handled\n", attributes);
+        if (access)
+            FIXME("access %#x not handled\n", access);
+        if (name)
+            FIXME("name %s not handled\n", debugstr_w(name));
+
+        win32_handle_info.sType = VK_STRUCTURE_TYPE_MEMORY_GET_WIN32_HANDLE_INFO_KHR;
+        win32_handle_info.pNext = NULL;
+        win32_handle_info.memory = resource->mem.device_allocation.vk_memory;
+        win32_handle_info.handleType = VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT;
+
+        vr = VK_CALL(vkGetMemoryWin32HandleKHR(device->vk_device, &win32_handle_info, handle));
+
+        ID3D12Resource_Release(resource_iface);
+        return vr ? E_FAIL : S_OK;
+    }
+
+    FIXME("Creating shared handle for type of object %p unsupported\n", object);
     return E_NOTIMPL;
 }
 
+static inline bool handle_is_kmt_style(HANDLE handle)
+{
+    return (ULONG_PTR)handle & 0x40000000 && ((ULONG_PTR)handle - 2) % 4 == 0;
+}
+
 static HRESULT STDMETHODCALLTYPE d3d12_device_OpenSharedHandle(d3d12_device_iface *iface,
         HANDLE handle, REFIID riid, void **object)
 {
-    FIXME("iface %p, handle %p, riid %s, object %p stub!\n",
+    struct d3d12_device *device = impl_from_ID3D12Device(iface);
+    HRESULT hr;
+
+    TRACE("iface %p, handle %p, riid %s, object %p\n",
             iface, handle, debugstr_guid(riid), object);
 
+    if (IsEqualGUID(riid, &IID_ID3D12Resource))
```

Comment:
What about ID3D12Resource1?

Reply:
Yeah just forgot, I'll add that

Reply:
Actually, looking at the documentation for ::OpenSharedHandle, it only mentions the ability to pass in the GUIDs for the base interfaces, I guess if it matters we could add a test to see whether this was just an oversight in the documentation.

https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-opensharedhandle

Reply:
We can always fix it later if need be.